pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/LiteralTransformFunction.java

```diff
@@ -128,64 +122,88 @@ public Dictionary getDictionary() {
 
   @Override
   public int[] transformToIntValuesSV(ProjectionBlock projectionBlock) {
-    if (_intResult == null) {
-      _intResult = new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    int numDocs = projectionBlock.getNumDocs();
+    Object ref = _result;
+    if (!(ref instanceof int[]) || ((int[]) ref).length < numDocs) {
```

Comment:
Per the current way of query execution, we can make the following assumptions:
1. Only one type of the result will be read
2. The first projection block will contain the most `numDocs` (up to 10K)

With these assumptions we can change the check to
```suggestion
    if (ref == null) {
      ...
    } else {
      assert (ref instanceof int[]) && ((int[]) ref).length >= numDocs;
    }
```

If the assumption break, we'd better let it throw exception, or the performance will degrade because we will end up filling the array multiple times.

Reply:
What if something changes? E.g. JSON often has variant types (e.g. string or number)

Reply:
instanceof checks aren’t worth trying to optimise away anyway, they’re intrinsified and speculatively eliminated with an uncommon trap, just like null checks

Reply:
The check itself is not causing the problem. The problem is that if the behavior changes, we might end up filling the array again and again, and causing performance degradation. E.g. if a function always read the literal as both INT and LONG for each block, we will keep creating new arrays and fill them.
So I'd suggest guarding this potential misuse of the function by throwing an exception.

Reply:
I'm ok with throwing an exception if the literal is used as the wrong type.

Reply:
Maybe this needs a different approach - one of these is created per segment, so we should construct all literals in the scope of the query context, then we don't care about the footprint of a single instance.

Reply:
Please take a look now, this change now only consists of the size change (plus allowing for racy writes while guaranteeing consistent reads) - the caching is done in #7720